tsl/src/compression/compression.c

```diff
@@ -2267,93 +2746,93 @@ decompress_batches_for_insert(const ChunkInsertState *cis, TupleTableSlot *slot)
 				 errmsg("inserting into compressed chunk with unique constraints disabled"),
 				 errhint("Set timescaledb.enable_dml_decompression to TRUE.")));
 
-	Oid comp_relid = ts_chunk_get_relid(cis->compressed_chunk_id, false);
-	Relation in_rel = relation_open(comp_relid, RowExclusiveLock);
+	Assert(OidIsValid(cis->compressed_chunk_table_id));
+	Relation in_rel = relation_open(cis->compressed_chunk_table_id, RowExclusiveLock);
+	CompressionSettings *settings = ts_compression_settings_get(cis->compressed_chunk_table_id);
+	Assert(settings);
 
-	RowDecompressor decompressor = build_decompressor(in_rel, out_rel);
 	Bitmapset *key_columns = compressed_insert_key_columns(out_rel);
 	Bitmapset *null_columns = NULL;
-
-	int num_scankeys;
-	ScanKeyData *scankeys = build_scankeys(cis->hypertable_relid,
-										   in_rel->rd_id,
-										   &decompressor,
-										   key_columns,
-										   &null_columns,
-										   slot,
-										   &num_scankeys);
-
-	bms_free(key_columns);
-
-	/*
-	 * Using latest snapshot to scan the heap since we are doing this to build
-	 * the index on the uncompressed chunks in order to do speculative insertion
-	 * which is always built from all tuples (even in higher levels of isolation).
-	 */
-	TupleTableSlot *compressed_slot = table_slot_create(in_rel, NULL);
-	Snapshot snapshot = GetLatestSnapshot();
-	TableScanDesc scan = table_beginscan(in_rel, snapshot, num_scankeys, scankeys);
-
-	while (table_scan_getnextslot(scan, ForwardScanDirection, compressed_slot))
+	struct decompress_batches_stats stats;
+
+	int num_index_scankeys;
+	Relation index_rel = NULL;
+	ScanKeyData *index_scankeys = build_index_scankeys_using_slot(cis->hypertable_relid,
+																  in_rel,
+																  out_rel,
+																  settings,
+																  slot,
+																  &index_rel,
+																  &num_index_scankeys);
+
+	if (index_rel)
 	{
-		bool valid = true;
-
-		/*
-		 * Since the heap scan API does not support SK_SEARCHNULL we have to check
-		 * for NULL values manually when those are part of the constraints.
-		 */
-		for (int attno = bms_next_member(null_columns, -1); attno >= 0;
-			 attno = bms_next_member(null_columns, attno))
+		/* Prepare the heap scan keys if any */
+		Bitmapset *filtered_key_columns = NULL;
+		int i = -1;
+		while ((i = bms_next_member(key_columns, i)) > 0)
 		{
-			if (!slot_attisnull(compressed_slot, attno))
+			AttrNumber attno = i + FirstLowInvalidHeapAttributeNumber;
+			char *attname = get_attname(out_rel->rd_id, attno, false);
+			if (!ts_array_is_member(settings->fd.segmentby, attname))
 			{
-				valid = false;
-				break;
+				filtered_key_columns = bms_add_member(filtered_key_columns, i);
```

Comment:
Here you're filtering out all segmentby columns because they should be handled by an index? Let's add a comment about this.

This is going to break when we allow indexes that cover a subset of segmentby columns, so might be good to start using a bitmap of columns handled by the index instead.